asyncio_pipe/__init__.py

```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env python
+# coding: utf8
+
+""" Asynchronous multiprocessing communication. """
+
+import asyncio
+
+from multiprocessing import Pipe as SyncPipe
+from multiprocessing.connection import Connection as SyncConnection
+from typing import Any, Optional, Tuple
+
+
+class Connection(object):
+    """
+        Asynchronous wrapper over multiprocessing.Connection class.
+        Register reader and write over given connection handle to
+        perform asynchronous I/O.
+
+        Parameters:
+            sync_connection (multiprocessing.Connection):
+                Synchronous process connection to decorate with
+                asynchronous methods.
+            loop (Optional[asyncio.AbstractEventLoop]):
+                Event loop to register connection I/O to, if `None``
+                then `asyncio.get_event_loop()` will be used, default
+                to `None`.
+    """
+
+    def __init__(
+            self,
+            sync_connection: SyncConnection,
+            loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
+        self._loop: asyncio.AbstractEventLoop
+        if loop is None:
+            self._loop = asyncio.get_event_loop()
+        else:
+            self._loop = loop
+        self._sync_connection: SyncConnection = sync_connection
+        self._read_event: asyncio.Event = asyncio.Event()
+        self._write_event: asyncio.Event = asyncio.Event()
+        self._loop.add_reader(
+            self._sync_connection.fileno(),
+            self._read_event.set)
+        self._loop.add_writer(
+            self._sync_connection.fileno(),
+            self._write_event.set)
+
+    def __del__(self) -> None:
+        self.close()
+
+    def fileno(self):
+        return self._sync_connection.fileno()
+
+    def close(self) -> None:
+        self._loop.remove_reader(self.fileno())
+        self._loop.remove_writer(self.fileno())
+        self._sync_connection.close()
+
+    @property
+    def closed(self) -> bool:
+        """True if the connection is closed"""
+        return self._sync_connection.closed
+
+    @property
+    def readable(self) -> bool:
+        """True if the connection is readable"""
+        return self._sync_connection.readable
+
+    @property
+    def writable(self) -> bool:
+        """True if the connection is writable"""
+        return self._sync_connection.writable
+
+    async def send(self, obj: Any) -> None:
+        """
+            Send a picklable object asynchronously.
+
+            See also:
+                multiprocessing.Connection#send
+        """
+        await self._write_event.wait()
```

Comment:
The [docs](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.add_writer) say "invoke callback with the specified arguments once fd is available for writing." It seems like this call to wait would block the first or second time send gets called. This line would require something to call `self._write_event.set` to be called before each send call. I don't see what would be doing this.